src/cursor/cur_hs.c

```diff
@@ -511,78 +509,33 @@ __curhs_search_near(WT_CURSOR *cursor, int *exactp)
-                if (cmp >= 0)
-                    break;
-            }
-            /*
-             * No entries greater than or equal to the key we searched for. Reset cursor if we get
-             * WT_NOTFOUND.
-             */
-            WT_ERR(ret);
-
-            *exactp = cmp;
+        WT_ERR_NOTFOUND_OK(__curhs_next_visible(session, hs_cursor), true);
+        if (ret == WT_NOTFOUND) {
+            WT_ERR(__wt_hs_cursor_prev(session, file_cursor));
+            WT_ERR(__curhs_prev_visible(session, hs_cursor));
         } else
-            *exactp = 1;
-
-        WT_ERR(__curhs_next_visible(session, hs_cursor));
+            WT_ERR(ret);
     }
-    /* Search the closest match that is smaller or equal to the search key. */
+    /*
+     * Walk backward first to try to increase the chance to place the cursor at the key that is
+     * smaller or equal to the search key.
+     */
     else {
-        /*
-         * Because of the special visibility rules for the history store, a new key can appear in
-         * between our search and the set of updates that we're interested in. Keep trying until we
-         * find it.
-         *
-         * There may be no history store entries for the given btree id and record key if they have
-         * been removed by rollback to stable.
-         *
-         * Note that we need to compare the raw key off the cursor to determine where we are in the
-         * history store as opposed to comparing the embedded data store key since the ordering is
-         * not guaranteed to be the same.
-         */
-        if (exact > 0) {
-            /*
-             * It's possible that we may race with a history store insert for another key. So we may
-             * be more than one record away the end of our target key/timestamp range. Keep
-             * iterating backwards until we land on our key.
-             */
-            while ((ret = __wt_hs_cursor_prev(session, file_cursor)) == 0) {
-                WT_STAT_CONN_INCR(session, cursor_skip_hs_cur_position);
-                WT_STAT_DATA_INCR(session, cursor_skip_hs_cur_position);
-
-                WT_ERR(__wt_compare(session, NULL, &file_cursor->key, srch_key, &cmp));
-                if (cmp <= 0)
-                    break;
-            }
-
-            *exactp = cmp;
+        WT_ERR_NOTFOUND_OK(__curhs_prev_visible(session, hs_cursor), true);
+        if (ret == WT_NOTFOUND) {
+            WT_ERR(__wt_hs_cursor_next(session, file_cursor));
+            WT_ERR(__curhs_next_visible(session, hs_cursor));
         } else
-            *exactp = -1;
-#ifdef HAVE_DIAGNOSTIC
-        if (ret == 0) {
-            WT_ERR(__wt_compare(session, NULL, &file_cursor->key, srch_key, &cmp));
-            WT_ASSERT(session, cmp <= 0);
-        }
-#endif
-
-        WT_ERR(__curhs_prev_visible(session, hs_cursor));
+            WT_ERR(ret);
     }
 
+    WT_ERR(__wt_compare(session, NULL, &file_cursor->key, srch_key, exactp));
```

Comment:
I can make an optimization to avoid doing this comparison if we haven't moved our cursor after calling `__wt_hs_cursor_search_near `. But the code is complex and not very clean with that optimization. I'll leave it aside unless someone has a clean way to do this.